tao_compiler/mlir/disc/transforms/shape_utils.cc

```diff
@@ -0,0 +1,1743 @@
+    auto input = op->getOperand(0);
+    auto in_ty = input.getType().dyn_cast_or_null<RankedTensorType>();
+    if (in_ty != nullptr && in_ty.getRank() == 1) {
+      updateShapeValueEqual(input, op->getResult(0));
+    }
+  }
+
+  return success();
+}
+
+LogicalResult ShapeAnalysis::buildOperationShapeMap(Operation* op) {
+  // build shapes for the results of op
+  for (Value result : op->getResults()) {
+    if (failed(buildSymbolShape(result))) {
+      return op->emitError("failed to build shape for op's result");
+    }
+  }
+
+  // collect shape tensor values and map equivalences.
+  buildOperationShapeValueMap(op);
+
+  // apply op's shape constraint
+  return applyOpConstraint(op);
+}
+
+LogicalResult ShapeAnalysis::buildRegionDimValueMap(Region* region) {
+  // Only SCF is supported a.t.m.
+  if (region->getBlocks().size() != 1) {
+    return region->getParentOp()->emitError(
+        "only single block region is supported");
+  }
+  for (Block& block : *region) {
+    if (failed(buildBlockDimValueMap(&block))) return failure();
+  }
+  applyDimEqualWithDimValue();
+  return success();
+}
+
+LogicalResult ShapeAnalysis::buildBlockDimValueMap(Block* block) {
+  // TODO: deal with memref.load and store directory rather than infer from
+  // reshape.
+  auto extractLmhloValueOfDim = [&](Value valueOfDims, int64_t index,
+                                    DimValue& dimValue) {
+    // case 1: dimValues are built with alloc and store.
+    //     %0 = memref.alloc() : memref<3xi32, "cpu">
+    //     memref.store %1, %0[%c0] : memref<3xi32, "cpu">
+    // TODO: there may be more scenarios.
+    for (auto user : valueOfDims.getUsers()) {
+      if (auto store = dyn_cast_or_null<memref::StoreOp>(user)) {
+        auto indices = store.getIndices();
+        if (indices.size() != 1) {
+          continue;
+        }
+        int64_t index_val;
+        if (getConstIntValue(indices[0].getDefiningOp(), index_val) &&
+            index_val == index) {
+          dimValue = DimValue(store.value());
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+  auto mayCreateOrMapConstInt = [&](DimValue& dimVal) {
+    DimValue intDimVal;
+    if (dimVal.state == DimValue::FOLD) {
+      intDimVal = dimVal;
+    } else if (dimVal.state == DimValue::UNFOLD) {
+      Operation* op = dimVal.unfoldVal.getDefiningOp();
+      if (isa<arith::ConstantIntOp, arith::ConstantIndexOp>(op)) {
+        int64_t result;
+        if (getConstIntValue(op, result) && result > 0) {
+          intDimVal = DimValue(result);
+        }
+      }
+    }
+    if (intDimVal.isValid()) {
+      if (dimVal.state == DimValue::UNFOLD) {
+        mapDimValueEqual(dimVal, intDimVal);
+      }
+    }
+  };
+
+  WalkResult result = block->walk([&](Operation* op) {
+    WalkResult res = WalkResult::advance();
+    // TODO: analysis memref::load/store.
+    if (auto dynamicReshape = dyn_cast<lmhlo::DynamicReshapeOp>(*op)) {
```

Comment:
So we only expect this phase to function in buffer level?
Maybe it's necessary to have an additional option is to avoid the overhead when the analysis is run on tensor level.